lib/lambda_ethereum_consensus/validator/validator_set.ex

```diff
@@ -0,0 +1,284 @@
+    end
+  end
+
+  defp update_validators(new_validators, set), do: %{set | validators: new_validators}
+
+  ##############################
+  # Attestation
+
+  defp attests(set, epoch, slot, head_root) do
+    case Duties.current_attesters(set.duties, epoch, slot) do
+      [] ->
+        set
+
+      attester_duties ->
+        attester_duties
+        |> Enum.map(&attest(&1, slot, head_root, set.validators))
+        |> update_duties(set, epoch, :attesters, slot)
+    end
+  end
+
+  defp publish_aggregates(set, epoch, slot) do
+    case Duties.current_aggregators(set.duties, epoch, slot) do
+      [] ->
+        set
+
+      aggregator_duties ->
+        aggregator_duties
+        |> Enum.map(&publish_aggregate(&1, slot, set.validators))
+        |> update_duties(set, epoch, :attesters, slot)
+    end
+  end
+
+  defp attest(duty, slot, head_root, validators) do
+    validators
+    |> Map.get(duty.validator_index)
+    |> Validator.attest(duty, slot, head_root)
+
+    Duties.attested(duty)
+  end
+
+  defp publish_aggregate(duty, slot, validators) do
+    validators
+    |> Map.get(duty.validator_index)
+    |> Validator.publish_aggregate(duty, slot)
+
+    Duties.aggregated(duty)
+  end
+
+  defp update_duties(new_duties, set, epoch, kind, slot) do
+    set.duties
+    |> Duties.update_duties!(kind, epoch, slot, new_duties)
+    |> then(&%{set | duties: &1})
+  end
+
+  ##############################
+  # Key management
+
+  @doc """
+    Get validator keystores from the keystore directory.
+    This function expects two files for each validator:
+      - <keystore_dir>/<public_key>.json
+      - <keystore_pass_dir>/<public_key>.txt
+  """
+  @spec decode_validator_keystores(binary(), binary()) ::
+          list(Keystore.t())
+  def decode_validator_keystores(keystore_dir, keystore_pass_dir)
+      when is_binary(keystore_dir) and is_binary(keystore_pass_dir) do
+    keystore_dir
+    |> File.ls!()
+    |> Enum.flat_map(&paths_from_filename(keystore_dir, keystore_pass_dir, &1, Path.extname(&1)))
+    |> Enum.flat_map(&decode_key/1)
+  end
+
+  defp paths_from_filename(keystore_dir, keystore_pass_dir, filename, ".json") do
+    basename = Path.basename(filename, ".json")
+
+    keystore_file = Path.join(keystore_dir, "#{basename}.json")
+    keystore_pass_file = Path.join(keystore_pass_dir, "#{basename}.txt")
+
+    [{keystore_file, keystore_pass_file}]
+  end
+
+  defp paths_from_filename(_keystore_dir, _keystore_pass_dir, basename, _ext) do
+    Logger.warning("[Validator] Skipping file: #{basename}. Not a json keystore file.")
+    []
+  end
+
+  defp decode_key({keystore_file, keystore_pass_file}) do
+    # TODO: remove `try` and handle errors properly
+    [Keystore.decode_from_files!(keystore_file, keystore_pass_file)]
+  rescue
+    error ->
+      Logger.error(
+        "[Validator] Failed to decode keystore file: #{keystore_file}. Pass file: #{keystore_pass_file} Error: #{inspect(error)}"
+      )
+
+      []
```

Comment:
Nit: aren't this functions more related to the Keystore module?

Reply:
They are! The Keystore PR was merged in the middle of this one and this was still in the previous `Setup` module instad of the Keystore so I mantained it here, but I agree with you, I created a small issue to move this and also create an `add` and `delete` in the ValidatorSet module. #1281